c3.php

```diff
@@ -8,420 +8,463 @@
 
-if (isset($_COOKIE['CODECEPTION_CODECOVERAGE'])) {
-    $cookie = json_decode($_COOKIE['CODECEPTION_CODECOVERAGE'], true);
+// $_SERVER['HTTP_X_CODECEPTION_CODECOVERAGE_DEBUG'] = 1;
 
-    // fix for improperly encoded JSON in Code Coverage cookie with WebDriver.
-    // @see https://github.com/Codeception/Codeception/issues/874
-    if (!is_array($cookie)) {
-        $cookie = json_decode($cookie, true);
-    }
+class C3 {
 
-    if ($cookie) {
-        foreach ($cookie as $key => $value) {
-            if (!empty($value)) {
-                $_SERVER["HTTP_X_CODECEPTION_" . strtoupper($key)] = $value;
-            }
-        }
-    }
-}
+    /**
+     * @var string
+     */
+    private $configDir;
 
-if (!array_key_exists('HTTP_X_CODECEPTION_CODECOVERAGE', $_SERVER)) {
-    return;
-}
+    /**
+     * @var string
+     */
+    private $autoloadRootDir;
 
-if (!function_exists('__c3_error')) {
-    function __c3_error($message)
+    /**
+     * @param string $configDir where to look for codeception.yml and codeception.dist.yml
+     */
+    public function __construct($configDir)
     {
-        $errorLogFile = defined('C3_CODECOVERAGE_ERROR_LOG_FILE') ?
-            C3_CODECOVERAGE_ERROR_LOG_FILE :
-            C3_CODECOVERAGE_MEDIATE_STORAGE . DIRECTORY_SEPARATOR . 'error.txt';
-        if (is_writable($errorLogFile)) {
-            file_put_contents($errorLogFile, $message);
-        } else {
-            $message = "Could not write error to log file ($errorLogFile), original message: $message";
-        }
-        if (!headers_sent()) {
-            header('X-Codeception-CodeCoverage-Error: ' . str_replace("\n", ' ', $message), true, 500);
-            setcookie('CODECEPTION_CODECOVERAGE_ERROR', $message);
-        }
+        $this->configDir = realpath($configDir);
+        $this->autoloadRootDir = $this->configDir;
     }
-}
 
-// Autoload Codeception classes
-if (!class_exists('\\Codeception\\Codecept') || !function_exists('codecept_is_path_absolute')) {
-    if (file_exists(__DIR__ . '/codecept.phar')) {
-        require_once 'phar://' . __DIR__ . '/codecept.phar/autoload.php';
-    } elseif (stream_resolve_include_path(__DIR__ . '/vendor/autoload.php')) {
-        require_once __DIR__ . '/vendor/autoload.php';
-        // Required to load some methods only available at codeception/autoload.php
-        if (stream_resolve_include_path(__DIR__ . '/vendor/codeception/codeception/autoload.php')) {
-            require_once __DIR__ . '/vendor/codeception/codeception/autoload.php';
-        }
-    } elseif (stream_resolve_include_path('Codeception/autoload.php')) {
-        require_once 'Codeception/autoload.php';
-    } else {
-        __c3_error('Codeception is not loaded. Please check that either PHAR or Composer package can be used');
+    /**
+     * @param string $dir directory from where to attempt autoloading Codeception
+     */
+    public function setAutoloadRootDir($dir)
+    {
+        $this->autoloadRootDir = $dir;
```

Comment:
I don't think the autoloadRootDir should be mutable. I'm not even sure it's needed at all in this class.